### DISCUSSION TITLE
Error in downloadBitmap - java.io.IOException: Unable to connect to: file://data/data/

### DISCUSSION THREAD

**Discussion by AAlex-11** [NativeScript/NativeScript#10535] [General]
I define iconRef in Angular Nativescript code as:

    <Image #icon />
    @ViewChild("icon", { static: false }) iconRef: ElementRef;

and try dynamically changing image as

    this.iconRef.nativeElement.src=`file:/${this.getfullIconPath(this.iconPath)}`

file .png is present, I'm sure that file present and path is correct. I can see file on Android studio

![JavaError-05012024_232655_1](https://github.com/NativeScript/NativeScript/assets/138773344/8a8cb4e5-061b-49db-b5e4-d204296efabd)

However I receive error

     Error in downloadBitmap - java.io.IOException: Unable to connect to: file://data/data/...
     Missing Image with resourceID: file://data/data/...

What going wrong?

**Comment by rigor789** [comment]
Where are you placing these icons?

Have you tried putting them in the `src/assets/` folder and using `~/assets/icons/${name}.png` as the path?

**Comment by AAlex-11** [comment]
@rigor789 , thanks for answer, this images created dynamically. Of course, static images I placed to src/assets/
Dynamically created image I placed by the same rules

     getfullIconPath(iconName:string){
        const appRootFolder = knownFolders.currentApp();
        const iconPath = path.join(appRootFolder.path, "icons")
        const iconFolder: Folder = Folder.fromPath(iconPath);
        return path.join(iconFolder.path, iconName);
     }

icons (images) created successfully, I can see they in the same place, can open it with Android studio and check what exactly I created.
But, of course, than I want to upload my images to Image control, but failed. Why? What going wrong? 

![Images-05022024_215157-1](https://github.com/NativeScript/NativeScript/assets/138773344/12d02819-1113-4632-b934-1369523db4f1)

**Comment by AAlex-11** [comment]
try to use 

        this.iconRef.nativeElement.src=`~/icons/1.png`

and receive error, however in reality image present, can opened and checked by Android Studio

![Images-05022024_220058-1](https://github.com/NativeScript/NativeScript/assets/138773344/d3862b74-a88f-4481-b13e-76991867339b)

**Comment by rigor789** [comment]
Not quite sure, I would try generating them in `assets/gen/icons` for example:
![image](https://github.com/NativeScript/NativeScript/assets/879060/d3faaa31-bb25-4db9-85ee-cec649c3a7a7)

And then reference them as `~/assets/gen/icons/${name}.png` or whatever. 

Thinking about it, maybe you don't need the `file:/` prefix just an absolute path and that might also work:
```ts
this.iconRef.nativeElement.src=this.getfullIconPath(this.iconPath)}`
```

**Comment by AAlex-11** [comment]
If I use static predefinition preloading image from source code ~/assets all working fine and icon showed correctly

       this.iconRef.nativeElement.src="~/assets/save1.png"

if I use dynamically created icon

       this.iconRef.nativeElement.src=this.iconPath

 it not working at all, despite icon is present, file is correct and can be opened by Android Studio

![Nativescript-05032024_235859_1](https://github.com/NativeScript/NativeScript/assets/138773344/78fdfe12-2f1c-4040-90aa-7b1df0f83a66)

this is my code in debugger, I simple replaced preloading image from source code to my dynamically created image:

![Nativescript-05042024_001337_1](https://github.com/NativeScript/NativeScript/assets/138773344/8bdd1707-e990-4052-9b2c-e64c58a84256)

And application crashed.  I try to add point to refresh and one more than

     .then((x) => (this.iconRef.nativeElement.src = `${this.iconPath}${"?ts=" + new Date().getTime()}`))

but nothing changing

![Nativescript-05042024_003943-1](https://github.com/NativeScript/NativeScript/assets/138773344/22d64244-077e-4dc4-804e-d122a94b72b8)

**Comment by rigor789** [comment]
The code screenshot there looks quite odd with that `new Promise` and `resolve` this way. What's `this.fileService.writeIcon` doing?

**Comment by AAlex-11** [comment]
write image to ~/assets/icon   
separate task (Promise) no matter, I can delete it, nothing changing after that, Image can not reading and uploading to Image control

**Reply by rigor789** [reply]
You are likely incorrectly awaiting the promises (so it does matter!) and trying to display an image before it is written to the filesystem...

Here's a working example: https://stackblitz.com/edit/nativescript-stackblitz-templates-dbdzom?file=src%2Fapp%2Fitem%2Fitems.component.html,src%2Fapp%2Fitem%2Fitems.component.ts&title=NativeScript%20Starter%20Angular

https://github.com/NativeScript/NativeScript/assets/879060/c57dc1e9-0833-4edb-9e72-201f68c790ac

![image](https://github.com/NativeScript/NativeScript/assets/879060/7fd7184b-4865-4a5a-b7aa-6430672855b7)

**Reply by AAlex-11** [reply]
thank you, maybe you are right
